core/optaplanner-core-impl/src/main/java/org/optaplanner/core/impl/solver/ConsumerSupport.java

```diff
@@ -97,15 +108,20 @@ private void tryConsumeWaitingIntermediateBestSolution() {
      */
     private CompletableFuture<Void> scheduleIntermediateBestSolutionConsumption() {
         return CompletableFuture.runAsync(() -> {
-            Solution_ bestSolution = bestSolutionWaitingForConsumption.getAndSet(null);
-            try {
-                if (bestSolution != null) {
-                    bestSolutionConsumer.accept(bestSolution);
+            BestSolutionContainingProblemChanges<Solution_> bestSolutionContainingProblemChanges = bestSolutionHolder.take();
+            if (bestSolutionContainingProblemChanges != null) {
+                try {
+                    bestSolutionConsumer.accept(bestSolutionContainingProblemChanges.getBestSolution());
+                } catch (Throwable throwable) {
+                    exceptionHandler.accept(problemId, throwable);
+                } finally {
+                    /*
+                     * Complete the CompletableFutures associated with ProblemChanges regardless of a possible
+                     * exception in the consumer to avoid a memory leak.
+                     */
+                    bestSolutionContainingProblemChanges.completeProblemChanges();
```

Comment:
Shouldn't those futures be completed exceptionally in case of an exception?

Reply:
Hmmm. You are right that the user can request the exception handler. But is the user *required* to? (And if not, what happens if he does not provide one? Is the exception just swallowed?)

Reply:
No, if the exception handler is not provided, a default one is used, which just logs the exception. Not ideal, though.

Reply:
Then maybe we should complete exceptionally in that case? Or is that too complicated already?

Reply:
In the end, I decided to complete the future exceptionally in the case described above. If the user is waiting for the best solution, it's safer to make sure the exception in the consumer will not be missed.